cpp/src/io/parquet/experimental/page_index_filter.cu

```diff
@@ -0,0 +1,779 @@
+
+    // Gather page-level string chars to row-level string chars
+    cudf::detail::batched_memcpy_async(src_iter, dst_iter, size_iter, total_rows, stream);
+
+    // Return row-level (output) strings children and the nullmask
+    return std::tuple{
+      std::move(row_str_chars), std::move(row_str_offsets), std::move(output_nullmask)};
+  }
+
+  /**
+   * @brief Builds two device columns storing the corresponding page-level statistics (min, max)
+   *        respectively of a column at each row index.
+   *
+   * @tparam T underlying type of the column
+   * @param schema_idx Column schema index
+   * @param dtype Column data type
+   * @param stream CUDA stream
+   * @param mr Device memory resource
+   *
+   * @return A pair of device columns with min and max value from page statistics for each row
+   */
+  template <typename T>
+  std::pair<std::unique_ptr<column>, std::unique_ptr<column>> operator()(
+    cudf::size_type schema_idx,
+    cudf::data_type dtype,
+    rmm::cuda_stream_view stream,
+    rmm::device_async_resource_ref mr) const
+  {
+    // List, Struct, Dictionary types are not supported
+    if constexpr (cudf::is_compound<T>() and not cuda::std::is_same_v<T, string_view>) {
+      CUDF_FAIL("Compound types other than strings do not have statistics");
+    } else {
+      // Compute column chunk level page count offsets, and page level row counts and row offsets.
+      auto const [page_row_counts, page_row_offsets, col_chunk_page_offsets] =
+        make_page_row_counts_and_offsets(per_file_metadata, row_group_indices, schema_idx, stream);
+
+      CUDF_EXPECTS(
+        page_row_offsets.back() == total_rows,
+        "The number of rows must be equal across row groups and pages within row groups");
+
+      auto const total_pages = col_chunk_page_offsets.back();
+
+      // Create host columns with page-level min, max values
+      host_column<T> min(total_pages, stream);
+      host_column<T> max(total_pages, stream);
+
+      // Populate the host columns with page-level min, max statistics from the page index
+      auto page_offset_idx = 0;
+      // For all row data sources
+      std::for_each(
+        thrust::counting_iterator<size_t>(0),
+        thrust::counting_iterator(row_group_indices.size()),
+        [&, col_chunk_page_offsets = col_chunk_page_offsets](auto src_idx) {
+          // For all column chunks in this source
+          auto const& rg_indices = row_group_indices[src_idx];
+          std::for_each(rg_indices.cbegin(), rg_indices.cend(), [&](auto rg_idx) {
+            auto const& row_group = per_file_metadata[src_idx].row_groups[rg_idx];
+            auto col              = std::find_if(
+              row_group.columns.begin(),
+              row_group.columns.end(),
+              [schema_idx](ColumnChunk const& col) { return col.schema_idx == schema_idx; });
+
+            // No need to check column_index and offset_index again
+            if (col != std::end(row_group.columns)) {
```

Comment:
when is a column(?) not found in `per_file_metadata`?

Reply:
Copy pasted from the row group version of this code segment, added a CUDF_EXPECTS instead.